### ISSUE TITLE
General permission methods

### ISSUE THREAD

**Issue by maxceem** [topcoder-platform/tc-project-service#346]
During working on the Workstreams we had to check permissions based on `allowRule` and `denyRule` as was described in this comment https://github.com/topcoder-platform/tc-project-service/issues/315#issuecomment-503396151. And it was implemented during the challenge in [workManagementForTemplate](https://github.com/topcoder-platform/tc-project-service/pull/342/files#diff-b4c12893f152cab9852018540be29a4d) permission middleware.

As we are going to have more and more places where we would have to restrict permissions based on various set of rules and to avoid issues like this one https://github.com/topcoder-platform/tc-project-service/issues/332 I've moved the logic for checking permissions into general util methods:
- `matchPermissionRule`
- `hasPermission`
- `hasPermissionForProject`

You can check them out in the PR https://github.com/topcoder-platform/tc-project-service/pull/345

The most general method is `hasPermissionForProject: (permission, user, projectId)`:
- `projectId` is project id we should check permission for
- `user` is the user we should check permissions for, usually we would use the current user from the request object `user = req.authUser`
- `permission`

I suggest to be able to define the `permission` in two ways:
- **Full** when permission is defined with `allowRule` and `denyRule`:

    ```js
    {
      allowRule: {
          projectRoles: [],
          topcoderRoles: []
      },
      denyRule: {
          projectRoles: [],
          topcoderRoles: []
      }
    }
    ```
- **Simplified** if we don't need `denyRule` we can deifne directly the inner part of `allowRule` without explicit property `allowRule`:

   ```js
   {
      projectRoles: [],
      topcoderRoles: []
   }
   ```
   This **simplified** permission is equal to a **full** permission:
    ```js
    {
       allowRule: {
         projectRoles: [],
         topcoderRoles: []
       }
    }
    ```

I guess this should cover most of our needs for permissions checking. And after some time we can refactor exitent code to use these methods instead of reimplementing this logic, for example many of [these methods](https://github.com/topcoder-platform/tc-project-service/tree/dev/src/permissions) could be rewritten reusing the general methods.

So far what I see can be done to improve them:
1. Create some way to check that the user is a member of the project with any member role. For now we would have to list all the user roles in `projectRoles` which is too wordy, also when adding a new role we would have to update all permission rules. So far I see two possible ways:
   1. Add new property like `isProjectMember` so the rule would look like:
      ```js
       {
         isProjectMember: true,
         topcoderRoles: []
       }
      ```
       The big disadvantage I see, is that it may be misleading as it could feel that we have to add both propeties `isProjectMember` and `projectRoles` to check if user is a member and has a particular role.
   2. Another way - is somehow use `projectRoles` property to define all roles at once. Possible ways:
       1. `projectRoles = []` - empty array, feels like areally bad idea
       2. `projectRoles = true` - boolean `true` is nice, the only small disadvantage I see, the word `true` doesn't exmplains what it means
      3. `projectRoels = 'any'` - string `any`, is good that it explains what does it mean, but it feels a little bit "hardcoded". We can introduce a new constant `ANY = 'any'`, but then we would be required to always import this constant which feels not that comfortable.
      4. Maybe you have any other idea.
 
2. cover these methods with extensive unit tests
3. use Joi to validate possible shapes of `permission` to provide detailed feedback to the developer about possible method misusing.
4. Write wiki documentation on how to use these methods with examples.

@vikasrohit Let me know what you think. If you see any ways we can adjust these methods to cover more cases or to be more comfortable or reliable.

**Comment by vikasrohit** [comment]
Agreed with all the points and plan to improve permissions overall. I have suggestion for specifying the permission rule for user being a project member without being too wordy. We can have a constant `PROJECT_MEMBER_ROLES` and populate it with all possible member roles once and then use this constants in the permission configuration.

**Comment by maxceem** [comment]
> I have suggestion for specifying the permission rule for user being a project member without being too wordy. 

You mean instead of [suggested](https://github.com/topcoder-platform/tc-project-service/commit/5a2070dec06249efe9a76ea3b94c4496700be3db#diff-c5e4a6f179e879e027c003a8daaf5f79R13) `ROLES_PROJECT_MEMBERS ` use `PROJECT_MEMBER_ROLES `?

**Comment by vikasrohit** [comment]
No, I meant:
we have all project member roles listed [here](https://github.com/topcoder-platform/tc-project-service/blob/5a2070dec06249efe9a76ea3b94c4496700be3db/src/constants.js#L16) so we can create our permissions as follows:
```
{
   projectRoles: _.values(PROJECT_MEMBER_ROLE),
   topcoderRoles: []
 }
```
So, we don't need any new config field for saying all project members and we don't have to specify/update the project member roles every where.

**Comment by maxceem** [comment]
Got you. The reason for such option like `projectRoles: true` is that we can specify permissions not only in the code config file, but also in the DB records. 

We already have several places where we specify permissions inside DB:
- ProjectService.writePermission
- ProjectService.readPermission
- WorkManagmentPermission.permission

So if define inside DB value as:
```
{
   projectRoles: ['manager', 'observer', 'customer', 'copilot', 'account_manager'],
   topcoderRoles: []
 }
```

Later, if we add a new project role we would have to update all the records in the DB.

**Comment by vikasrohit** [comment]
Hmm, I think we should keep it simple for now and specify each role in `projectRoles` and `topcoderRoles`. We can optimize things later.
